From 082de687f22e5daf37bf277c1c3c715c457a5c29 Mon Sep 17 00:00:00 2001
From: Shuang Liang <liangshuang@eswincomputing.com>
Date: Tue, 9 Jul 2024 09:38:51 +0800
Subject: [PATCH 41/69] drivers: usb: fix getting hub descriptor fail.

Fix the issue of hub not being able to read descriptor.

Signed-off-by: Shuang Liang <liangshuang@eswincomputing.com>
Signed-off-by: Pinkesh Vaghela <pinkesh.vaghela@einfochips.com>

(cherry picked from commit 86ede17f5023f2efada1e29bec963ecde93b89c8)
https://github.com/sifiveinc/riscv-linux/commit/86ede17f5023f2efada1e29bec963ecde93b89c8
---
 drivers/usb/core/hub.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 256fe8c86828..3651e03c5f6d 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -1464,6 +1464,7 @@ static int hub_configure(struct usb_hub *hub,
 	unsigned unit_load;
 	unsigned full_load;
 	unsigned maxchild;
+	int try_cnt = 0;
 
 	hub->buffer = kmalloc(sizeof(*hub->buffer), GFP_KERNEL);
 	if (!hub->buffer) {
@@ -1484,6 +1485,7 @@ static int hub_configure(struct usb_hub *hub,
 		goto fail;
 	}
 
+retry:
 	/* Request the entire hub descriptor.
 	 * hub->descriptor can handle USB_MAXCHILDREN ports,
 	 * but a (non-SS) hub can/will return fewer bytes here.
@@ -1503,9 +1505,15 @@ static int hub_configure(struct usb_hub *hub,
 		ret = -ENODEV;
 		goto fail;
 	} else if (hub->descriptor->bNbrPorts == 0) {
-		message = "hub doesn't have any ports!";
-		ret = -ENODEV;
-		goto fail;
+		try_cnt++;
+		if (try_cnt < 10) {
+			mdelay(10);
+			goto retry;
+		} else {
+			message = "hub doesn't have any ports!";
+			ret = -ENODEV;
+			goto fail;
+                }
 	}
 
 	/*
-- 
2.51.0

